packages/twenty-server/src/engine/modules/calendar/timeline-calendar-event.service.ts

```diff
@@ -0,0 +1,266 @@
+          workspaceMemberId: attendee.workspaceMemberId,
+          firstName,
+          lastName,
+          displayName,
+          avatarUrl,
+          handle: attendee.handle,
+        };
+      });
+
+    const calendarEventAttendeesByEventId: {
+      [calendarEventId: string]: TimelineCalendarEventAttendee[];
+    } = groupBy(formattedCalendarEventAttendees, 'calendarEventId');
+
+    const totalNumberOfCalendarEvents: { count: number }[] =
+      await this.workspaceDataSourceService.executeRawQuery(
+        `
+      SELECT
+          COUNT(DISTINCT "calendarEventId")
+      FROM
+          ${dataSourceSchema}."calendarEventAttendee" "calendarEventAttendee"
+      WHERE
+          "calendarEventAttendee"."personId" = ANY($1)
+      `,
+        [personIds],
+        workspaceId,
+      );
+
+    const timelineCalendarEvents = calendarEvents.map((event) => {
+      const attendees = calendarEventAttendeesByEventId[event.id] || [];
+
+      return {
+        ...event,
+        attendees,
+      };
+    });
+
+    const calendarEventIdsWithWorkspaceMemberInAttendees =
+      await this.workspaceDataSourceService.executeRawQuery(
+        `
+      SELECT
+          "calendarEventId"
+      FROM
+          ${dataSourceSchema}."calendarEventAttendee" "calendarEventAttendee"
+      WHERE
+          "calendarEventAttendee"."workspaceMemberId" = $1
+      `,
+        [workspaceMemberId],
+        workspaceId,
+      );
+
+    const calendarEventIdsWithWorkspaceMemberInAttendeesFormatted =
+      calendarEventIdsWithWorkspaceMemberInAttendees.map(
+        (event: { calendarEventId: string }) => event.calendarEventId,
+      );
+
+    const calendarEventIdsToFetchVisibilityFor = timelineCalendarEvents
+      .filter(
+        (event) =>
+          !calendarEventIdsWithWorkspaceMemberInAttendeesFormatted.includes(
+            event.id,
+          ),
+      )
+      .map((event) => event.id);
+
+    const calendarEventIdsForWhichVisibilityIsMetadata:
+      | {
+          id: string;
+        }[]
+      | undefined = await this.workspaceDataSourceService.executeRawQuery(
+      `
+      SELECT
+          "calendarChannelEventAssociation"."calendarEventId" AS "id"
+      FROM
+          ${dataSourceSchema}."calendarChannel" "calendarChannel"
+      LEFT JOIN
+          ${dataSourceSchema}."calendarChannelEventAssociation" "calendarChannelEventAssociation" ON "calendarChannel".id = "calendarChannelEventAssociation"."calendarChannelId"
+      WHERE
+          "calendarChannelEventAssociation"."calendarEventId" = ANY($1)
+      AND
+          "calendarChannel"."visibility" = 'METADATA'
+      `,
+      [calendarEventIdsToFetchVisibilityFor],
+      workspaceId,
+    );
+
+    if (!calendarEventIdsForWhichVisibilityIsMetadata) {
+      throw new Error('Failed to fetch calendar event visibility');
+    }
+
+    timelineCalendarEvents.forEach((event) => {
+      event.visibility = calendarEventIdsForWhichVisibilityIsMetadata.some(
```

Comment:
for better performances, you can create a map from calendarEventIdsForWhichVisibilityIsMetadata and then have quick acces with calendarEventIdsForWhichVisibilityIsMetadataMap.get(event.id)